tensorstore/kvstore/s3/s3_credential_provider.h

```diff
@@ -0,0 +1,112 @@
+
+#ifndef TENSORSTORE_KVSTORE_S3_S3_CREDENTIAL_PROVIDER_H_
+#define TENSORSTORE_KVSTORE_S3_S3_CREDENTIAL_PROVIDER_H
+
+#include <functional>
+#include <string>
+#include <string_view>
+#include <map>
+#include <memory>
+
+#include "absl/synchronization/mutex.h"
+#include "absl/time/time.h"
+#include "tensorstore/internal/http/curl_handle.h"
+#include "tensorstore/internal/http/curl_transport.h"
+#include "tensorstore/internal/http/http_transport.h"
+#include "tensorstore/util/result.h"
+
+using ::tensorstore::Result;
+
+namespace tensorstore {
+namespace internal_auth_s3 {
+
+
+class S3Credentials {
+private:
+ std::string access_key_;
+ std::string secret_key_;
+ std::string session_token_;
+
+public:
+  explicit S3Credentials(std::string_view access_key="",
+                         std::string_view secret_key="",
+                         std::string_view session_token="")
+    : access_key_(access_key),
+      secret_key_(secret_key),
+      session_token_(session_token)
+    {}
+
+  void SetAccessKey(std::string_view access_key) { access_key_ = access_key; }
+  void SetSecretKey(std::string_view secret_key) { secret_key_ = secret_key; }
+  void SetSessionToken(std::string_view session_token) { session_token_ = session_token; }
+
+  const std::string & GetAccessKey() const { return access_key_; }
+  const std::string & GetSecretKey() const { return secret_key_; }
+  const std::string & GetSessionToken() const { return session_token_; }
+
+  bool IsValid() const { return !access_key_.empty(); }
+};
+
+class CredentialProvider {
+ public:
+  virtual ~CredentialProvider() = default;
+  virtual Result<S3Credentials> GetCredentials() = 0;
+};
+
+class EnvironmentCredentialProvider : public CredentialProvider {
+ private:
+  S3Credentials credentials_;
+ public:
+  EnvironmentCredentialProvider(const S3Credentials & credentials)
+    : credentials_(credentials) {}
+  virtual Result<S3Credentials> GetCredentials() override
+    { return credentials_; }
+};
+
+class FileCredentialProvider : public CredentialProvider {
+ private:
+  absl::Mutex mutex_;
+  std::string filename_;
+  std::string profile_;
+ public:
+  FileCredentialProvider(std::string_view filename, std::string_view profile) :
+    filename_(filename), profile_(profile) {}
+  virtual Result<S3Credentials> GetCredentials() override;
+};
+
+class EC2MetadataCredentialProvider : public CredentialProvider {
+ private:
+  std::shared_ptr<internal_http::HttpTransport> transport_;
+ public:
+  EC2MetadataCredentialProvider(std::shared_ptr<internal_http::HttpTransport> transport)
+    : transport_(transport) {}
+  virtual Result<S3Credentials> GetCredentials() override
+    { return absl::UnimplementedError("EC2 Metadata Server"); }
+};
+
+using S3CredentialProvider =
+    std::function<Result<std::unique_ptr<CredentialProvider>>()>;
+
+void RegisterS3CredentialProviderProvider(S3CredentialProvider provider, int priority);
+
+Result<std::unique_ptr<CredentialProvider>> GetS3CredentialProvider(
```

Comment:
The comments for this method should indicate the order in which credentials are tried.  i.e file, environment variables, etc.

Reply:
commented